### PR TITLE
Add script for automatic validation

### PR DESCRIPTION
**Pull request by ZanezZephyrs**
Not sure this should be in this repository, but i think it is better to leave it here for safekeeping for now.
This is the script to automatic download, convert and evaluate the t5x checkpoints, it is meant to be run daily in a cronjob later.

### CODE REVIEW

File: t5x/scripts/download_and_evaluate_checkpoints.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--config_path", default=None, required=True, help="path for a config file, see the sample config in the repository")
+    parser.add_argument("--path_for_lm_eval", default=None, required=False, help="local path for lm eval repository")
+    parser.add_argument("--path_for_t5x_to_hf", default=None, required=False, help="path for the directory containing the script")
+
+    args = parser.parse_args()
+
+    
+    PATH_FOR_LM_EVAL=args.path_for_lm_eval or "/home/thales/lm-evaluation-harness"
```
Comment (rodrigonogueira4) [inline t5x/scripts/download_and_evaluate_checkpoints.py:81]: nit: remove hard coded paths?

File: t5x/scripts/download_and_evaluate_checkpoints.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
+# script to download, convert and evaluate checkpoints
```
Comment (rodrigonogueira4) [inline t5x/scripts/download_and_evaluate_checkpoints.py:1]: nit: capitalize "Script" and add a period at the end sentence?

File: t5x/scripts/download_and_evaluate_checkpoints.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+    
+    PATH_FOR_LM_EVAL=args.path_for_lm_eval or "/home/thales/lm-evaluation-harness"
+    PATH_FOR_T5X_TO_HF=args.path_for_t5x_to_hf or "/home/thales/run_eval/t5x/t5x/scripts/"
+
+    CONFIG_FILE=args.config_path
+    config=json.load(open(CONFIG_FILE))
+    
+
+    # PATH="rodrigo-gpt-j-test-temp-11/checkpoint_"
```
Comment (hugoabonizio) [inline t5x/scripts/download_and_evaluate_checkpoints.py:88]:
```suggestion
```

**Review by hugoabonizio — APPROVED** [review]

File: t5x/scripts/download_and_evaluate_checkpoints.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+    if config.get("calculate_mc4_loss", None):
+        print("CALCULATING MC4 LOSS")
+        results_save_file=Path(results_save_dir,f"mc4_loss_results.csv")
+
+        if os.path.isfile(results_save_file):
+            print(f"WARNING: the file {results_save_file} already exists, this may lead to error in retrieving metrics")
+            _ = input("Press enter to continue anyway")  
+
+        base_command_for_mc4_loss_calculation=f"cd {PATH_FOR_MC4_LOSS_SCRIPT} && python3 calculate_mc4_loss.py"
```
Comment (rodrigonogueira4) [inline t5x/scripts/download_and_evaluate_checkpoints.py:171]: Do we have this script already in the repo?

**Comment by rodrigonogueira4** [comment]
Thanks a lot for doing this. Could you please run it through a lint so we start with the correct formatting?

File: t5x/scripts/download_and_evaluate_checkpoints.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        mc4_loss_command=f"{base_command_for_mc4_loss_calculation} --model_path {converted_checkpoint_save_path} --output_path {results_save_file} --device {config['DEVICE']}"
+
+        result=run(mc4_loss_command,shell=True)
+
+        mc4_loss_results=pd.read_csv(results_save_file)
+        mc4_metrics=["custom_loss", "hf_loss"]
```
Comment (rodrigonogueira4) [inline t5x/scripts/download_and_evaluate_checkpoints.py:177]: remove `hf_loss` and rename `custom loss` to `cross_entropy_loss`?

File: t5x/scripts/download_and_evaluate_checkpoints.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+        wandb.init(project=config["wandb_project"], entity=config["wandb_entity"], id=config["EXPERIMENT_LABEL"], resume="allow")
+        wandb.config = {
+            "checkpoint": selected_checkpoint,
+            "bucket": config["BUCKET_NAME"],
+            "Prefix": config["PATH"],
+        }
+
+
+    #download t5x
```
Comment (rodrigonogueira4) [inline t5x/scripts/download_and_evaluate_checkpoints.py:125]: nit: in PEP8, comments should start capitalized, end with a period, and have a space between the hashtag and the comment itself:
`# Download T5x`

File: t5x/scripts/convert_gptj_t5x_to_hf.py [hunk 7]
```diff
@@ -9,14 +9,27 @@
...
 import torch
+from pathlib import Path
+import json
+from tqdm import tqdm
+
+def patch_config_file(hf_checkpoint_path):
+    config_path=Path(hf_checkpoint_path,"config.json")
+    config_data=json.load(open(config_path))
+
+    if "layer_norm_epsilon" in config_data:
```
[the remark this answers is not in the document]
Comment (ZanezZephyrs) [reply t5x/scripts/convert_gptj_t5x_to_hf.py]: yeah, i guess we don need it

File: t5x/scripts/download_and_evaluate_checkpoints.py (hunk 4 above)
Comment (ZanezZephyrs) [reply t5x/scripts/download_and_evaluate_checkpoints.py]: i added the script now

File: t5x/scripts/calculate_mc4_loss.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    parser.add_argument(
+        "--device",
+        default="cuda",
+        required=False,
+        help="local path for lm eval repository",
+    )
+
+    args = parser.parse_args()
+
+    batch_size = 4
```
Comment (rodrigonogueira4) [inline t5x/scripts/calculate_mc4_loss.py:40]: add these as arguments?

File: t5x/scripts/convert_gptj_t5x_to_hf.py [hunk 9]
```diff
@@ -3,126 +3,170 @@
...
+    out_file = open(config_path, "w")
+    json.dump(config_data, out_file, indent=4)
+
+
+def convert_t5x_checkpoint_to_flax(
+    t5x_checkpoint_path, config_name, hf_checkpoint_path
+):
     config = GPTJConfig.from_pretrained(config_name)
     gptj_model = GPTJForCausalLM(config=config)
     t5x_model = checkpoints.load_t5x_checkpoint(t5x_checkpoint_path)
```
Comment (rodrigonogueira4) [inline t5x/scripts/convert_gptj_t5x_to_hf.py:33]: maybe print a message saying "Loading T5x checkpoint"?

**Comment by rodrigonogueira4** [comment]
Thanks for implementing this!

**Review by rodrigonogueira4 — APPROVED** [review]

**Review by piresramon — APPROVED** [review]
